src/mlpack/core/data/dataset_info_impl.hpp

```diff
@@ -14,56 +14,82 @@ namespace mlpack {
-  // we create a mapping.
-  if (maps.count(dimension) == 0 ||
-      maps[dimension].first.left.count(string) == 0)
-  {
-    // This string does not exist yet.
-    size_t& numMappings = maps[dimension].second;
-    if (numMappings == 0)
-      types[dimension] = Datatype::categorical;
-    typedef boost::bimap<std::string, size_t>::value_type PairType;
-    maps[dimension].first.insert(PairType(string, numMappings));
-    return numMappings++;
-  }
-  else
-  {
-    // This string already exists in the mapping.
-    return maps[dimension].first.left.at(string);
-  }
+  // Nothing to initialize here.
+}
+
+// When we want to insert value into the map,
+// we could use the policy to map the string
+template<typename PolicyType>
+inline typename PolicyType::mapped_type DatasetMapper<PolicyType>::MapString(
+                                                    const std::string& string,
+                                                    const size_t dimension)
+{
+  return policy.template MapString<MapType>(string, dimension, maps, types);
 }
 
 // Return the string corresponding to a value in a given dimension.
-inline const std::string& DatasetInfo::UnmapString(
-    const size_t value,
-    const size_t dimension)
+template<typename PolicyType>
+inline const std::string& DatasetMapper<PolicyType>::UnmapString(
+                                                    const size_t value,
+                                                    const size_t dimension)
 {
   // Throw an exception if the value doesn't exist.
   if (maps[dimension].first.right.count(value) == 0)
   {
     std::ostringstream oss;
-    oss << "DatasetInfo::UnmapString(): value '" << value << "' unknown for "
-        << "dimension " << dimension;
+    oss << "DatasetMapper<PolicyType>::UnmapString(): value '" << value
+        << "' unknown for dimension " << dimension;
     throw std::invalid_argument(oss.str());
   }
 
   return maps[dimension].first.right.at(value);
 }
 
+// Return the value corresponding to a string in a given dimension.
+template<typename PolicyType>
+inline typename PolicyType::mapped_type DatasetMapper<PolicyType>::UnmapValue(
+                                                    const std::string& string,
+                                                    const size_t dimension)
+{
+  // Throw an exception if the value doesn't exist.
+  if (maps[dimension].first.left.count(string) == 0)
+  {
+    std::ostringstream oss;
+    oss << "DatasetMapper<PolicyType>::UnmapValue(): string '" << string
+        << "' unknown for dimension " << dimension;
+    throw std::invalid_argument(oss.str());
+  }
+
+  return maps[dimension].first.left.at(string);
+}
+
+template<typename PolicyType>
+template<typename eT>
+inline void DatasetMapper<PolicyType>::MapTokens(
+                                        const std::vector<std::string>& tokens,
+                                        size_t& row,
+                                        arma::Mat<eT>& matrix)
```

Comment:
The tabbing here seems weird, if you put the arguments on a new line, they should only be two tabs in from the original indentation.  I think this comment applies to some other methods in this PR too.